Add a nice header to our README.md (#123093)
rust-lang/rust: README.md

<div align="center">

>   <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg" width="512" height="256">

Reviewer: This doesn't need to all be on one line like it currently is, how about
```suggestion
  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software"
    src="https://www.rust-lang.org/static/images/rust-social-wide.jpg"
    width="512" height="256">
```

Maybe the other way around? :thinking:

> [Website](https://www.rust-lang.org/) | [Getting started](https://www.rust-lang.org/learn/get-started) | [Learn](https://www.rust-lang.org/learn) | [Documentation](https://www.rust-lang.org/learn#learn-use) | [Contributing](CONTRIBUTING.md)

Reviewer: That's better!

I think this should either have multiple lines still, separated by the pipe char, or move the links into the "link block" at the end of this section. Whatever you think would be nicer for the inherent split of visualizing the result + diffs, ultimately.

Author: Since we are already using "link block" in other parts of the README.md, I decided to use those (and did the same for the tools).

</div>

This is the main source code repository for [Rust]. It contains the compiler,
standard library, and documentation.

[Rust]: https://www.rust-lang.org/

## Why Rust?

- **Performance:** Fast and memory-efficient, suitable for critical services, embedded devices, and easily integrate with other languages.

- **Reliability:** Our rich type system and ownership model ensure memory and thread safety, reducing bugs at compile-time.

> - **Productivity:** Comprehensive documentation, user-focused compiler, and advanced tooling including package manager and build tool ([Cargo](https://github.com/rust-lang/cargo)), auto-formatter ([rustfmt](https://github.com/rust-lang/rustfmt)), linter ([Clippy](https://github.com/rust-lang/rust-clippy)) and editor support ([rust-analyzer](https://github.com/rust-lang/rust-analyzer)).

Reviewer: Hm I think we can express it like...
```suggestion
- **Productivity:** Comprehensive documentation, a compiler committed to providing great diagnostics, and advanced tooling including package manager and build tool ([Cargo](https://github.com/rust-lang/cargo)), auto-formatter ([rustfmt](https://github.com/rust-lang/rustfmt)), linter ([Clippy](https://github.com/rust-lang/rust-clippy)) and editor support ([rust-analyzer](https://github.com/rust-lang/rust-analyzer)).
```

Author: Sure, I like this. Done.

> <p align="center">

Reviewer: If this all looks the same with either the `header` tag or the `section` tag those would be preferable. Also the `align` attribute is deprecated, so assuming it renders the same please write this as
```suggestion
<header style="text-align:center;">
```

Author: This is sadly not possible, as GitHub removes any inline style.

We also can't use `<header>` or `<section>` since neither of them supports the `align="center"` hack.

And we can't continue to use `<p>` since it doesn't work with real Markdown links, so I changed it to `<div align="center">`.

Reviewer: I "love" this discovery. Apologies, I tried it out in a CommonMark demo before recommending it and didn't realize the [GFM spec](https://github.github.com/gfm/) was out of date.

  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg" width="512" height="256">

>   <br>

Reviewer: Per my last comment you probably don't even need this, just a space.

  <a href="https://www.rust-lang.org/">Website</a> |
  <a href="https://www.rust-lang.org/learn/get-started">Getting started</a> |
  <a href="https://www.rust-lang.org/learn">Learn</a> |
  <a href="https://www.rust-lang.org/learn#learn-use">Documentation</a> |

>   <a href="CONTRIBUTING.md">Contributing</a>

Reviewer: These can continue to be normal Markdown links and thus all as one line, inline, with the links provided in Markdown "footer" later, making it clear how it should look. As long as you use line breaks appropriately between things, you can intermingle HTML and Markdown syntax in a CommonMark or GitHub-Flavored Markdown document (because GHF strives to be compatible with CommonMark). So just use a line break before and after the Markdown syntax. The HTML will still be included directly and will still affect the Markdown after.

Author: Done.

> </p>

Reviewer: ```suggestion
</header>
```

This is the main source code repository for [Rust]. It contains the compiler,
standard library, and documentation.

[Rust]: https://www.rust-lang.org/

## Why Rust?

- **Performance:** Fast and memory-efficient, suitable for critical services, embedded devices, and easily integrate with other languages.

- **Reliability:** Our rich type system and ownership model ensure memory and thread safety, reducing bugs at compile-time.

> 

Reviewer: I don't think these spaces between bullets are necessary.

Author: They change the readability of the bullet points. Without spaces I think it's too close together.

Without spaces:
![image](https://github.com/rust-lang/rust/assets/3616612/b2c033b5-dece-486e-8afd-ecacf5c03b3d)

With spaces:
![image](https://github.com/rust-lang/rust/assets/3616612/1a93a5ac-f7f1-4d18-b5cd-230e7ee44778)

Reviewer: Hmm, okay.

> - **Productivity:** Comprehensive documentation, user-friendly compiler, and advanced tooling including package manager, build tool, auto-formatter, and editor support.

Reviewer: This deviation from the website's way of expressing it is again driven by "this is the README for the compiler's source repo, so we should use it to introduce people to what they can find in this repo or related repos".

> - **Productivity:** Comprehensive documentation, user-friendly compiler, and advanced tooling including package manager, build tool, auto-formatter, and editor support.

Author: > I feel like our compiler is slightly more-precise than "user-friendly": we want good diagnostics, and we treat bad diagnostics as bugs if they are at all improvable.

Completely agree, but I don't know how to say that with only 2/3 words. I therefor changed it to "user-focused" instead of "user-friendly" to emphasise the fact that we care about the users.

> I also think we can go ahead and cross-link here: we can identify that we ship this stuff via Rustup, and we can name cargo, rustfmt, rust-analyzer, and what they are, with hyperlinks.

Also agree, I've added a hyperlink to all the major tools we ship.

<div align="center">
  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software"
       src="https://www.rust-lang.org/static/images/rust-social-wide.jpg"

>        width="512" height="256">

Reviewer: I think we can fix the dark mode thing as a followup maybe but I checked with a couple people and am getting "logo too big" feedback. It's especially annoying on smaller devices, please try the [srcset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#srcset) attribute as a fix?

Reviewer: Alternatively twiddling the `width` as a % may be better?

Author: What about `width="50%"`?

![Capture d’écran 2024-04-01 à 21 38 15](https://github.com/rust-lang/rust/assets/3616612/fd64e51b-3542-4085-a78f-9cfbabb3a5ec)
![Screenshot 2024-04-01 at 21-38-32 Urgau_rust at improve-readme](https://github.com/rust-lang/rust/assets/3616612/9409d2c9-668e-4d23-a9a8-744ed25bbb2d)

<p align="center">

>   <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg">

Reviewer 2: imo this image is way too big, it's not very useful and just takes away space.

Author: yeah, the social image from rust-lang.org has unfortunately the right ratio to be very big; I will try to reduce it's size

Author: done.

  <br>
  <a href="https://www.rust-lang.org/">Website</a> |
  <a href="https://www.rust-lang.org/learn/get-started">Getting started</a> |
  <a href="https://www.rust-lang.org/learn">Learn</a> |
  <a href="https://www.rust-lang.org/learn#learn-use">Documentation</a> |
  <a href="CONTRIBUTING.md">Contributing</a>
</p>

This is the main source code repository for [Rust]. It contains the compiler,
standard library, and documentation.

[Rust]: https://www.rust-lang.org/

> ## Why Rust?

Reviewer 2: i'm also not sure whether it's worth having this. i don't expect that people visiting this repository need to be convinced that rust is good, i would expect that people coming to the repository are rust users already. maybe i'm wrong though.

Author: maybe, maybe not; I would expect to have at least some people who don't know Rust at all, to come to our README first before anything else, and giving them (without indirection) reason to like the language is IMO only positive

Reviewer: I don't disagree with _having_ the Why but we can keep it very terse and make fewer unsubstantiated / overly-subjective claims. Some people I've met claim Rust _doesn't_ have great documentation, because they are used to ecosystems which are even more thoroughly documented (though I disagree about "great"). However we do have e.g. integrated doc tooling.

> ## Why Rust?

Author: @workingjubilee in https://github.com/rust-lang/rust/pull/123093#discussion_r1539751987

> I don't disagree with _having_ the Why but we can keep it very terse and make fewer unsubstantiated / overly-subjective claims. Some people I've met claim Rust _doesn't_ have great documentation, because they are used to ecosystems which are even more thoroughly documented (though I disagree about "great" with them, they are allowed to have their own criteria). However we do have e.g. integrated doc tooling.

Author: These statements were taken literally from the website, so let's edit them at source first. Otherwise, it will be a bit of a disavowal of them.

Reviewer: I think it's fine if those are written in a different tone because it's not on the primary README for the distribution of the program which should first and foremost serve those who need information about this motley bundle of files.

Author: Feel free to suggest modifications/improvements.

<details>
<summary>LLM improved version</summary>

### Performance
Rust is known for its high speed and efficient memory usage, making it suitable for performance-intensive tasks such as powering critical services, running on embedded devices, and seamless integration with other programming languages.

### Reliability
Rust's robust type system and ownership model ensure memory and thread safety, reducing the likelihood of bugs at compile-time.

### Productivity
Rust offers comprehensive documentation, a user-friendly compiler with helpful error messages, and advanced tooling including an integrated package manager, build tool, multi-editor support with auto-completion and type inspections, an auto-formatter, and additional features.

</details>

Reviewer: Yeah hitting our points hard and fast seems like an improvement on the previous version, hmm.

### Performance

> Rust is blazingly fast and memory-efficient: with no runtime or garbage collector, it can power performance-critical services, run on embedded devices, and easily integrate with other languages.

Reviewer: If Rust doesn't have a runtime, how do you explain the runtime startup?

https://github.com/rust-lang/rust/blob/master/library/rtstartup/rsbegin.rs

We both know what people mean, but this is the surface of the documentation for the language, it needn't be written in a marketing-speak tone.

Author: There is no longer any mention of "runtime".
